### ISSUE TITLE
[FR] Support non-Gaussian likelihoods

### ISSUE THREAD

**Issue by tom-andersson** [tom-andersson/deepsensor#95]
### Summary

DeepSensor currently only supports Gaussian likelihoods with the core `ConvNP` model. However, many environmental variables are non-Gaussian, e.g. precipitation.

The [model.predict method](https://github.com/tom-andersson/deepsensor/blob/ba66c35dc1bf577b7c7e791bc9277f1a1512cfb1/deepsensor/model/model.py#L226) currently assumes a Gaussian likelihood. See some offending lines below: https://github.com/tom-andersson/deepsensor/blob/ba66c35dc1bf577b7c7e791bc9277f1a1512cfb1/deepsensor/model/model.py#L477-L484

But what if we want more than just the mean and stddev of the model's output distribution? What if our model has other distribution parameters, like the bernoulli-prob and Gamma parameters of a Bernoulli-Gamma distribution? We would need to add the lines above to include calls to other prediction parameters, like `model.bernoulli_prob`, `model.gamma_alpha`, and `model.gamma_beta`.

Another small sticking point:  The `Prediction` object output by `model.predict` is set up to support storing prediction parameters other than `mean` and `std`, but the following is currently hard-coded into the `__init__`: https://github.com/tom-andersson/deepsensor/blob/ba66c35dc1bf577b7c7e791bc9277f1a1512cfb1/deepsensor/model/pred.py#L62-L63

One design idea: each `DeepSensorModel` has a `pred_parameters: List[str] = ["mean", "stddev"]` attribute. In the `ConvNP` class, this attribute can be updated based on the `likelihood` parameter. Then, in `model.predict`, this list is used to determine which model methods to call and which parameters are added to the `Prediction` object.

Writing this in haste for now so that it's on people's radar and available as a discussion thread. I have not spent much time thinking about this.

cc @wesselb @patel-zeel 

### Basic Example

Ideally I would like to be able to do something like this (skipping some lines from the quickstart):
```
# Construct ConvNP with Bernoulli-Gamma likelihood for modelling precipitation
model = ConvNP(data_processor, task_loader, likelihood="cnp-bernoulli-gamma")

# Predict on new task with 50 context points and a dense grid of target points
test_task = task_loader("2014-12-31", context_sampling=50)
pred = model.predict(test_task, X_t=ds_raw)
```
```
>>> pred["air"]
<xarray.Dataset>
Dimensions:  (time: 1, lat: 25, lon: 53)
Coordinates:
  * time     (time) datetime64[ns] 2014-12-31
  * lat      (lat) float32 75.0 72.5 70.0 67.5 65.0 ... 25.0 22.5 20.0 17.5 15.0
  * lon      (lon) float32 200.0 202.5 205.0 207.5 ... 322.5 325.0 327.5 330.0
Data variables:
    mean     (time, lat, lon) float32 267.7 267.2 266.4 ... 297.5 297.8 297.9
    std      (time, lat, lon) float32 9.855 9.845 9.848 ... 1.356 1.36 1.487
    bernoulli_prob      (time, lat, lon) float32 blah blah blah
    gamma_alpha      (time, lat, lon) float32 blah blah blah
    gamma_beta      (time, lat, lon) float32 blah blah blah
```

With potentially more prediction parameters than the 5 listed above.

### Drawbacks

It will be fairly tricky to generalise `model.predict` for this while maintaining all existing functionality (e.g. sampling functionality).

### Unresolved questions

The Bernoulli likelihood mentioned above is actually not currently supported in the `ConvNP`, because it is not supported by the [`construct_convgnp` method from `neuralprocesses`](https://github.com/wesselb/neuralprocesses/blob/8f04edc9afd0f3214020994c80dd83c8297362be/neuralprocesses/architectures/convgnp.py#L97), which `ConvNP` wraps around.

To support a `cnp-bernoulli-gamma` likelihood we would need a `Gamma` class to be added to `neuralprocesses` and then the [`SpikesSlab` class](https://github.com/wesselb/neuralprocesses/blob/f20572ba480c1279ad5fb66dbb89cbc73a0171c7/neuralprocesses/dist/spikeslab.py#L16) would need to be extended with a `BernoulliGamma` child class. Perhaps an issue should be raised over at `neuralprocesses` @wesselb? 🙂 

In the meantime, `ConvNP` *does* support a `cnp-spikes-beta` likelihood, a distribution with spikes at 0 and 1 and a Beta distribution in (0, 1). This could be a good way to test any new non-Gaussian functionality.

**Comment by wesselb** [comment]
@tom-andersson If a `BernoulliGamma` likelihood would be desirable, then it would definitely be possible to implement it in `neuralprocesses`. :) This should be much simpler compared to `spikes-beta`. Is this something that you would like support for?

On the topic of extensions for `neuralprocess`, I think I am a little behind on features that I promised to implement, most notably support for TF graph mode. Is there anything else important that you are still missing?

**Comment by wesselb** [comment]
@tom-andersson Thanks. :) I will then move that `BernoulliGamma` the top of my priority list for `neuralprocesses`!

JAX support should certainly be possible. It might require a bit of thinking, but the backend agnostic design should allow one to add another backend. I'll move that onto the priority list too then.

**Comment by tom-andersson** [comment]
Okay, I have non-Gaussian likelihoods provisionally working in DeepSensor locally. Will update here when this feature is available publicly. Here's an example using a non-Gaussian mixture likelihood with deltas/spikes at 0 and 1 and a Beta distribution in (0, 1):
```
>>> # Vanilla DeepSensor quickstart code here...
>>> model = ConvNP(data_processor, task_loader, likelihood="cnp-spikes-beta")
>>> pred = model.predict(test_task, X_t=ds_raw, pred_params=["mean", "std", "beta_dist_alpha", "beta_dist_beta"])
>>> pred
{'air': <xarray.Dataset>
Dimensions:          (time: 1, lat: 25, lon: 53)
Coordinates:
  * lat              (lat) float32 75.0 72.5 70.0 67.5 ... 22.5 20.0 17.5 15.0
  * lon              (lon) float32 200.0 202.5 205.0 207.5 ... 325.0 327.5 330.0
  * time             (time) datetime64[ns] 2014-12-31
Data variables:
    mean             (time, lat, lon) float32 287.2 287.1 287.1 ... 287.1 287.1
    std              (time, lat, lon) float32 6.875 6.863 6.854 ... 6.861 6.866
    beta_dist_alpha  (time, lat, lon) float32 0.4944 0.4847 ... 0.5013 0.5024
    beta_dist_beta   (time, lat, lon) float32 0.5048 0.5084 ... 0.5085 0.5079}
```

Above, you can see the alpha and beta parameters of the Beta distribution component in the `Prediction` object.

Under the hood, the `pred_params` kwarg of `model.predict` dictates which low-level model methods get called and returned in the `Prediction` object.

**Comment by tom-andersson** [comment]
Hey @wesselb, I've added three new methods to the `ConvNP` class: `beta_dist_alpha`, `beta_dist_beta`, and `mixture_probs`, each returning parameters of a `neuralprocesses` distribution object using `dist.slab.alpha`, `dist.slab.beta`, and `dist.logprobs`, respectively.

1. I was wondering if the interface will be identical for the planned Bernoulli-Gamma likelihood in `neuralprocesses`? E.g. `dist.slab.alpha` and `dist.slab.beta` will return the [Gamma distribution](https://www.wikiwand.com/en/Gamma_distribution) alpha and beta params, respectively.
2. If `"spikes-beta"` is the `likelihood` kwarg in `construct_convgnp` for the mixture likelihood above, what will the kwarg be for the new Bernoulli-Gamma likelihood? I can then set this up in DeepSensor ahead of the new feature in `neuralprocesses`. I provisionally have `"bernoulli-gamma"` but perhaps you have another preference.

**Comment by wesselb** [comment]
Hey @tom-andersson! Whoa, this is super nice. :) Exciting developments!!

To answer your questions:

1. Yes, it should be exactly the same interface!

2. `"bernoulli-gamma"` sounds like an eminently sensible choice, so I will go for that. :)

**Comment by tom-andersson** [comment]
@wesselb great! no rush!

**Comment by emily-bodekerscientific** [comment]
Hi @wesselb, I am using DeepSensor to model various atmospheric variables, and am currently implementing precipitation. Having a Bernoulli-Gamma likelihood would be hugely helpful for this - I was wondering if this is likely to be implemented in `neuralprocesses`? Apologies if it is and I've missed it - I have updated to the latest version and looked through the repo but can't see it. Thanks, Emily

**Comment by wesselb** [comment]
Hey @emily-bodekerscientific!

Apologies for the delay. It's a busy and hectic time.

It's currently not implemented, but it would be very feasible to do. At the moment, I do not have any capacity for this, but some capacity should free up in one to two weeks. I'll keep you posted!

**Comment by emily-bodekerscientific** [comment]
@wesselb Hi, thanks so much for the response!  Not a rush at all, but it would be a feature I'd really like to take advantage of at some point if possible. Look forward to hearing from you!

**Comment by wesselb** [comment]
@emily-bodekerscientific, I have a bit more space now. I'll start working on this over the next while. :) Can't make any guarantees about when this will be done, though!

**Comment by emily-bodekerscientific** [comment]
Thanks so much @wesselb, it's really appreciated!

**Comment by wesselb** [comment]
@emily-bodekerscientific, I just released a new version of `neuralprocesses` (`v0.2.6`) which supports the Bernoulli-gamma likelihood:

```
In [2]: import neuralprocesses.torch as nps
   ...: import torch
   ...:
   ...: model = nps.construct_convgnp(likelihood="bernoulli-gamma")
   ...:
   ...: model(
   ...:     torch.randn(1, 1, 10),
   ...:     torch.randn(1, 1, 10),
   ...:     torch.randn(1, 1, 10),
   ...: ).sample()

Out[2]:
tensor([[[0.1590, 0.2837, 0.0096, 0.7695, 0.0000, 0.0000, 0.0000, 1.2632,
          0.0000, 0.0000]]], grad_fn=<AddBackward0>)
```

It's unit tested, but still very new, so please open an issue if you run into any issues.

CC @tom-andersson

**Comment by emily-bodekerscientific** [comment]
Thanks so much @wesselb, I'll let you know how I get on with it! Hopefully a publication coming out of this work so I'll definitely put you and `neuralprocesses` in the acknowledgements.

**Comment by wesselb** [comment]
@emily-bodekerscientific Fantastic! :) Should you have any more specific questions or issues, please also feel free to reach out over email.

**Comment by tom-andersson** [comment]
@wesselb thanks very much for implementing the Bernoulli-Gamma likelihood!

I just tested `likelihood="bernoulli-gamma"` in DeepSensor and it almost works out-the-box. The only issue is that we need to support normalising data to [0, new_max_val] in DeepSensor. See snippet below and in particular the TEMP comment:

```
import deepsensor.torch
from deepsensor.data import DataProcessor, TaskLoader
from deepsensor.data.sources import get_era5_reanalysis_data
from deepsensor.model import ConvNP
from deepsensor.train import Trainer

import xarray as xr
import pandas as pd
import numpy as np
from tqdm import tqdm

# If we want to use the GPU, uncomment:
# deepsensor.train.set_gpu_default_device()

# Load raw data
extent = "uk"
data_range = ("2015-01-01", "2015-01-31")
era5_var_IDs = ["total_precipitation"]
ds_raw = get_era5_reanalysis_data(
    era5_var_IDs, extent, date_range=data_range, verbose=True)

# Some very small negative ERA5 precip values are possible
ds_raw['total_precipitation'] = ds_raw['total_precipitation'].where(ds_raw['total_precipitation'] > 0.0, 0.0)

# Normalise data
data_processor = DataProcessor(x1_name="lat", x2_name="lon")
ds = data_processor(ds_raw, method="min_max")

# TEMP: "min_max" normalisation maps data to [-1, 1], but `neuralprocesses` assumes a spike at 0
# We need to make the new min/max values in the DataProcessor configurable so we can set new_min=0
ds = ds + 1

# Set up task loader
task_loader = TaskLoader(context=ds, target=ds)

# Set up model
model = ConvNP(data_processor, task_loader, likelihood="bernoulli-gamma")

# Generate training tasks with up 100 grid cells as context and all grid cells
#   as targets
train_tasks = []
for date in ds.time.values:
    N_context = np.random.randint(0, 100)
    task = task_loader(date, context_sampling=N_context, target_sampling="all")
    train_tasks.append(task)

# Train model
trainer = Trainer(model, lr=5e-5)
for epoch in tqdm(range(10)):
    batch_losses = trainer(train_tasks)

# Predict on new task with 50 context points and a dense grid of target points
test_task = task_loader(ds_raw.time.values[0], context_sampling=50)
pred = model.predict(test_task, X_t=ds_raw, pred_params=["mean", "std", "mixture_probs"])
```

Here is a plot of the probability of the first mixture component (which I think is the spike at 0, i.e. no rain):
<img width="617" alt="image" src="https://github.com/user-attachments/assets/84c98dfc-5e69-40b0-8283-b617daffa4a2">

cc @emily-bodekerscientific

**Comment by tom-andersson** [comment]
Minor note @wesselb, I think we agreed that `dist.slab.alpha/beta` would return the alpha/beta parameters of the CNP's Gamma distribution, but I'm not seeing these in the `dist.slab` attributes. Users may want to extract these to constructor plot their own Gamma distributions, so could be useful :-)

**Comment by wesselb** [comment]
@tom-andersson, currently the gamma distrubution is parametrised by the shape `k` and scale `scale`: `slab.k` and `slab.scale`. I opted for the shape–scale parametrisation since I believe it is the most common parametrisation of gamma distributions. :)

**Comment by tom-andersson** [comment]
@wesselb great, thanks for clarifying that. I'll raise an issue here to expose those attributes in DeepSensor.
